database/src/env.rs

```diff
@@ -222,6 +227,54 @@ where
     /// a table doesn't exist.
     fn open_db_rw<T: Table>(&self, tx_rw: &Rw) -> Result<impl DatabaseRw<T>, RuntimeError>;
 
+    /// TODO
+    ///
+    /// # Errors
+    /// TODO
+    fn open_tables(&self, tx_ro: &Ro) -> Result<impl Tables, RuntimeError> {
+        Ok((
+            EnvInner::open_db_ro::<BlockInfoV1s>(self, tx_ro)?,
+            EnvInner::open_db_ro::<BlockInfoV2s>(self, tx_ro)?,
+            EnvInner::open_db_ro::<BlockInfoV3s>(self, tx_ro)?,
+            EnvInner::open_db_ro::<BlockBlobs>(self, tx_ro)?,
+            EnvInner::open_db_ro::<BlockHeights>(self, tx_ro)?,
+            EnvInner::open_db_ro::<KeyImages>(self, tx_ro)?,
+            EnvInner::open_db_ro::<NumOutputs>(self, tx_ro)?,
+            EnvInner::open_db_ro::<PrunedTxBlobs>(self, tx_ro)?,
+            EnvInner::open_db_ro::<PrunableHashes>(self, tx_ro)?,
+            EnvInner::open_db_ro::<Outputs>(self, tx_ro)?,
+            EnvInner::open_db_ro::<PrunableTxBlobs>(self, tx_ro)?,
+            EnvInner::open_db_ro::<RctOutputs>(self, tx_ro)?,
+            EnvInner::open_db_ro::<TxIds>(self, tx_ro)?,
+            EnvInner::open_db_ro::<TxHeights>(self, tx_ro)?,
+            EnvInner::open_db_ro::<TxUnlockTime>(self, tx_ro)?,
+        ))
+    }
+
+    /// TODO
+    ///
+    /// # Errors
+    /// TODO
+    fn open_tables_mut(&self, tx_rw: &Rw) -> Result<impl TablesMut, RuntimeError> {
+        Ok((
+            EnvInner::open_db_rw::<BlockInfoV1s>(self, tx_rw)?,
+            EnvInner::open_db_rw::<BlockInfoV2s>(self, tx_rw)?,
+            EnvInner::open_db_rw::<BlockInfoV3s>(self, tx_rw)?,
+            EnvInner::open_db_rw::<BlockBlobs>(self, tx_rw)?,
+            EnvInner::open_db_rw::<BlockHeights>(self, tx_rw)?,
+            EnvInner::open_db_rw::<KeyImages>(self, tx_rw)?,
+            EnvInner::open_db_rw::<NumOutputs>(self, tx_rw)?,
+            EnvInner::open_db_rw::<PrunedTxBlobs>(self, tx_rw)?,
+            EnvInner::open_db_rw::<PrunableHashes>(self, tx_rw)?,
+            EnvInner::open_db_rw::<Outputs>(self, tx_rw)?,
+            EnvInner::open_db_rw::<PrunableTxBlobs>(self, tx_rw)?,
+            EnvInner::open_db_rw::<RctOutputs>(self, tx_rw)?,
+            EnvInner::open_db_rw::<TxIds>(self, tx_rw)?,
+            EnvInner::open_db_rw::<TxHeights>(self, tx_rw)?,
+            EnvInner::open_db_rw::<TxUnlockTime>(self, tx_rw)?,
+        ))
+    }
```

Comment:
@Boog900 is it this part that you wish to replace with a macro ?

Reply:
I want to replace the Ok parts here with a function call generated by a macro and the `Tables(Mut)` definition with a macro as well

Reply:
I don't understand what you mean by *function call generated by a macro*. Why would you want to add another function inside this one ?

Reply:
Because currently this is confusing to change if we were to add/ delete tables.

If instead this was all generated by a macro then it would be as simple as adding another argument/ removing an argument from the macro

Reply:
@Boog900 I think I understand - all `$table:ty`'s should be defined in some `output_table_types!()` which just outputs a list of all table types - and that should be used as input everywhere?

I'm not sure how to do with this `macro_rules!()` as the output must be valid syntax - this isn't allowed:
```rust
macro_rules! output_table_types {
    () => {
        crate::tables::BlockInfoV1s,
        crate::tables::BlockInfoV2s,
        crate::tables::BlockInfoV3s,
        [...]
    };
}
```
I'll try finding something that works - but the idea of passing around an `impl TablesMut` that basically is a struct that holds our tables is OK?